haskell.{compiler,packages}.microhs: init at 0.15.4.0

AlexandreTunstall opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This change adds microhs-based packagesets to the `haskell.packages` package sets.
For example, `haskell.packages.microhs.mtl` is a working microhs-compiled `mtl` package.

The microhs compilers added here replace the recently added `microhs` package.
Unlike the previous version, all new versions are built entirely from source by using Hugs to interpret the microhs source code.
cpphs was also booted from source in a similar way.

The goal is to allow Nixpkgs users to more easily experiment with using microhs, and perhaps one day boot GHC from source.
Note, however, that `nix-shell` (and related derivations) is completely untested.

A few patches were needed to make microhs and microcabal behave properly inside the Nix store, all of which I will seek to upstream.
The patch to make microhs Hugs-compatible is more of a hack, and I expect keeping microhs Hugs-compatible to be too troublesome to be worth upstreaming.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [X] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AlexandreTunstall committed:
haskell.{compiler,packages}.microhs: init

The package set is built entirely from source using the stdenv and Hugs.

This also replaces the recently added microhs package, which was built
from pre-generated compiler output.

Co-authored-by: sternenseemann <sternenseemann@systemli.org>

steeleduncan reviewed (commented):
Could I ask what the reasoning for all the revision-versioned microhs packages is?

I know it is done with GHC, but that is an outlier for nixpkgs. With e.g. `gcc`, `clang`, `go`, etc there are only packages for the major versions with (deliberately) backwards incompatible changes

steeleduncan reviewed (changes requested):
If this is a fix to another commit in this PR it should be merged with that commit (ie the initial `haskell.{compiler,packages}.microhs-*: init` commit)

AlexandreTunstall commented:
Since 0.14.17.0, `array` isn't included in microhs' `base` package.

Would that count as "a fix to another commit"? I'm not sure.

sternenseemann commented on pkgs/development/compilers/microhs/boot.nix:
```diff
@@ -0,0 +1,50 @@
+  buildPhase = ''
+    runHook preBuild
+    runHook postBuild
+  '';
```
```suggestion
  dontBuild = true;
```

sternenseemann commented on pkgs/development/haskell-modules/configuration-microhs.nix:
```diff
@@ -0,0 +1,104 @@
+  unix = self.unix_2_8_7_0;
+  xhtml = self.xhtml_3000_4_0_0;
+}
+// lib.optionalAttrs (lib.versionAtLeast super.ghc.version "0.14.17.0") {
```
Should probably rename this to `compiler` or `hc` at some point and add `ghc` and `mhs` aliases for convenience… Additional would of course be that a lot of conditional code already gates stuff on `ghc.version`…

AlexandreTunstall replied:
Someone recently brought to my attention the fact that `haskellPackages.ghc` is `buildPackages.ghc` (runs on `buildPlatform`) when it should be `ghc` (runs on `hostPlatform`). That's another argument in favour of refactoring, though perhaps less important due to how few packages depend on `ghc`.

Atemu commented on pkgs/development/compilers/microhs/boot.nix:
```diff
@@ -0,0 +1,50 @@
+      printf ${lib.escapeShellArg ''
+        #!${runtimeShell}
+        s='%s/share/microhs-hugs/src'
+        exec ${hugs}/bin/runhugs "+P$s/hugs:$s/src:$s/paths:{Hugs}/packages/*:hugs/obj" -98 +o +w -h100m "$s/hugs/Main.hs" "$@"
```
If i'm reading this correctly, this is just a wrapper, right? Why make this so convoluted rather than just makeWrappering the runhugs binary?

Also: could we instead just build a binary using hugs? Please explain your reasoning in a code comment; no matter which way.

AlexandreTunstall replied:
I forgot that `makeWrapper` could prepend arguments. I will try it out.

> could we instead just build a binary using hugs?

I'm not sure I understand the question. Do you mean using a wrapped `runhugs` as the boot mhs instead of this compiled version?

Atemu replied:
I meant to build a mhs binary using hugs rather than interpreting it.

AlexandreTunstall replied:
The included combinator file has been compiled from the Haskell source files using MicroHs, which was itself presumably compiled by the maintainer using GHC or an older version of MicroHs. As a result, it is inherently less trustworthy than the readable source code, so we would rather not use it if we don't have to.

Atemu reviewed (commented):
If there is no specific need for other versions, I see no reason to keep them.

I propose we strive for the least amount of mhs compilers possible to complete bootstrap to the newest one.  
Ideally, this would be 1 but that's unrealistic. 

I could see a chain like this though:

0. hugs-compatible mhs ran/built via hugs
1. mhs that can build latest
2. latest mhs

AlexandreTunstall commented on pkgs/development/compilers/microhs/boot.nix:
```diff
@@ -0,0 +1,50 @@
+
+  inherit (microhs-src) src;
+
+  patches = [ patches/hugs.patch ];
```
The upstream hugs branch is not kept up-to-date.

We might be able to use the ancient version if we patch out all the newer primitives, but I expect that it would only get less compatible over time.

AlexandreTunstall commented on pkgs/development/compilers/microhs/boot.nix:
```diff
@@ -0,0 +1,50 @@
+
+  patches = [ patches/hugs.patch ];
+
+  buildPhase = ''
```
It removes the default `make` call. I will add a comment to the code to make that clear.

sternenseemann commented on pkgs/development/compilers/microhs/boot.nix:
```diff
@@ -0,0 +1,50 @@
+  doInstallCheck = true;
+
+  installCheckPhase = ''
+    $out/bin/mhs --version
```
Probably best to use `versionCheckHook`!

sternenseemann reviewed (commented):
Only a partial review, I won't have time to finish it today.

sternenseemann committed:
microhs: make alias/shortcut for haskell.compiler.microhs

sternenseemann committed:
top-level/release-haskell.nix: test some pkgs with MicroHs

sternenseemann reviewed (approved):
Overall LGTM

sternenseemann commented on pkgs/development/tools/haskell/microcabal/patches/nixpkgs-compat.patch:
As long as the PR is not merged it doesn't really matter.

AlexandreTunstall committed:
top-level/release-haskell.nix: test GHC boot pkgs

Test the GHC boot packages that are known to work with MicroHs and mark
those that don't work as broken.

sternenseemann committed:
haskell.packages.microhs: add doDistribute where necessary

Packages that have an empty hydraPlatforms won't be tested on Hydra
otherwise, even if explicitly requested.
